src/test/transport.rs

```diff
@@ -1,15 +1,102 @@
-//! Implementation of a transport for testing purposes. This is largely based on
-//! the `rust-web3` `TestTransport` type with some modifications.
+//! Implementation of a transport for testing purposes.
 
 use jsonrpc_core::{Call, Value};
 use std::cell::RefCell;
+use std::clone::Clone;
 use std::collections::VecDeque;
+use std::fmt::Debug;
 use std::rc::Rc;
 use web3::error::Error;
-use web3::futures::future::{self, FutureResult};
-use web3::helpers;
+use web3::futures::future::FutureResult;
 use web3::{RequestId, Transport};
 
+// ----------------------------------------------------
+
+/// Request
+#[derive(Debug)]
+pub struct Request {
+    pub method: String,
+    pub params: Vec<Value>,
+}
+
+/// RequestResponse
+#[derive(Debug)]
+pub struct RequestResponse {
+    request: Request,
+    response: Result<Value, Error>,
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct TestTransport_ {
+    request_responses: Rc<RefCell<VecDeque<RequestResponse>>>,
+}
+
+impl TestTransport_ {
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    pub fn expect_request(
+        &mut self,
+        method: impl Into<String>,
+        params: impl AsRef<[Value]>,
+        response: Result<Value, Error>,
+    ) {
+        self.request_responses
+            .borrow_mut()
+            .push_back(RequestResponse {
+                request: Request {
+                    method: method.into(),
+                    params: params.as_ref().to_vec(),
+                },
+                response,
+            });
+    }
+
+    pub fn assert_no_missing_requests(&self) {
+        if let Some(request_response) = self.request_responses.borrow().front() {
+            panic!(
+                "transport dropped without observing all expected requests: \
+                 first missing request is {}",
```

Comment:
nit: log params as well.

Reply:
slightly more idiomatic could be if this method returned an object that has a method `return(response: Result<Value, Error>)`

Then you could do something like

```rust
transport
  .given("eth_estimateGas", [json!({"to": to,"value": "0x2a"})])
  .return(Ok(json!("0x42")))
```

Also would it make sense to have params be optional?

Reply:
I really like this as it would make some of the test cases more legible:
```
transport
    .expect_request("eth_sendRawTransaction", [json!(tx_raw)])
    .with_response(Ok(json!(tx_hash)));
```

Reply:
I don't think its necessary because its just 3 parameters and after looking at one test it's clear what it means and it's more annoying to have to write with_response and the response is not optional.

But if you both prefer that version then I'll change it. Builder style we can also default to empty list for params  but the response is still not optional.

Reply:
I think that the request-response only gets added when `with_response` is called. We can also add a `#[must_use = "request expectations are ignored without a matching 'with_response()'"]` so you get a compiler warning when a `expect_response` is called without a matching `with_response`.

While I do agree that having to type out `with_response` is more characters, it does make reading the code easier IMO.

Compare the two versions:
```
transport.expect_request("eth_sendRawTransaction", [json!(tx_raw)], Ok(json!(tx_hash)))
transport
    .expect_request("eth_sendRawTransaction", [json!(tx_raw)])
    .with_response(Ok(json!(tx_hash)));
```

In the first, especially because the last two parameters both use `json!` macros, the request parameters and response JSON blur together in my eyes when reading.